### DISCUSSION TITLE
Get reference using a field other than the primary key

### DISCUSSION THREAD

**Discussion by haykerman** [mikro-orm/mikro-orm#4011] [Ideas]
Currently, we're able to get a reference to an entity using something like this:
`em.getReference<Book>(Book, id);`

But if I have a `uuid` column (not a primary key) defined on the Book entity, and my frontend <-> backend only exchange data with uuids, I am forced to query the database (`findOne({ uuid: "..." })`), to then use it to insert an Author (`.create({name: "Steven King", book })`)

It would be nice to support `getReference` using the `uuid` in this case.
`em.getReference<Book>(Book, { uuid: uuid });`

Not sure what query is generated using the primary key field and what should be generated using the non-primary-key field.

**Comment by B4nan** [comment]
You need to know the primary key, as that is what is used in foreign keys as well as in the identity map key. The example with `em.create` would just fire insert query, that wouldn't be possible if you don't provide the FK value.

Only way this could work is if we always load the entity during flush based on the unique property to get the PK value, it would also require changes to how the identity map indexes the entities, which would end up as a performance hit most probably - as to support composite PKs there would have to be non-trivial serialization support. So not entirely impossible, but much bigger thing than you might have guessed, I would say not worth it as it would do exactly the same as you already can with a single line of code. 

edit: Hmm scratch that, I don't think this would be possible to handle properly on identity map level, as the PK would need to work as well, and you could end up with multiple instances of a single entity, one created with PK and one not.

**Comment by haykerman** [comment]
Ah, got it. Makes sense. Thanks for the clarification and the awesome work
you're doing on the ORM.
